clippy_lints/src/unnecessary_reserve.rs

```diff
@@ -0,0 +1,150 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::{match_def_path, meets_msrv, msrvs, paths, visitors::expr_visitor_no_bodies};
+use rustc_errors::Applicability;
+use rustc_hir::{intravisit::Visitor, Block, ExprKind, QPath, StmtKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_semver::RustcVersion;
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// This lint checks for a call to `reserve` before `extend` on a `Vec` or `VecDeque`.
+    /// ### Why is this bad?
+    /// Since Rust 1.62, `extend` implicitly calls `reserve`
+    /// ### Example
+    /// ```rust
+    /// let mut vec: Vec<usize> = vec![];
+    /// let array: &[usize] = &[1, 2];
+    /// vec.reserve(array.len());
+    /// vec.extend(array);
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let mut vec: Vec<usize> = vec![];
+    /// let array: &[usize] = &[1, 2];
+    /// vec.extend(array);
+    /// ```
+    #[clippy::version = "1.64.0"]
+    pub UNNECESSARY_RESERVE,
+    pedantic,
+    "`reserve` before `extend` no longer makes sense in rustc version >= 1.62"
+}
+
+pub struct UnnecessaryReserve {
+    msrv: Option<RustcVersion>,
+}
+
+impl UnnecessaryReserve {
+    #[must_use]
+    pub fn new(msrv: Option<RustcVersion>) -> Self {
+        Self { msrv }
+    }
+}
+
+impl_lint_pass!(UnnecessaryReserve => [UNNECESSARY_RESERVE]);
+
+impl<'tcx> LateLintPass<'tcx> for UnnecessaryReserve {
+    fn check_block(&mut self, cx: &LateContext<'tcx>, block: &Block<'tcx>) {
+        if !meets_msrv(self.msrv, msrvs::UNNECESSARY_RESERVE) {
+            return;
+        }
+
+        for (idx, stmt) in block.stmts.iter().enumerate() {
+            if let StmtKind::Semi(semi_expr) = stmt.kind
+                && let ExprKind::MethodCall(_, [struct_calling_on, _], _) = semi_expr.kind
+                && let Some(expr_def_id) = cx.typeck_results().type_dependent_def_id(semi_expr.hir_id)
+                && (match_def_path(cx, expr_def_id, &paths::VEC_RESERVE) ||
+                    match_def_path(cx, expr_def_id, &paths::VEC_DEQUE_RESERVE))
+                && acceptable_type(cx, struct_calling_on)
+                && let Some(next_stmt_span) = check_extend_method(cx, block, idx, struct_calling_on)
+                && !next_stmt_span.from_expansion()
+            {
+                span_lint_and_then(
+                    cx,
+                    UNNECESSARY_RESERVE,
+                    next_stmt_span,
+                    "this `reserve` no longer makes sense in rustc version >= 1.62",
```

Comment:
```suggestion
                    "unnecessary call to `reserve`",
```

There should be a note explaining why the call is unnecessary as well, similar to the above comment.